File: java/src/main/java/com/aerospike/helper/collections/LargeList.java
package com.aerospike.helper.collections;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.aerospike.client.AerospikeClient;
import com.aerospike.client.Bin;
import com.aerospike.client.Key;
import com.aerospike.client.Record;
import com.aerospike.client.Value;
import com.aerospike.client.cdt.ListOperation;
import com.aerospike.client.policy.WritePolicy;
/**
 * An implementation of LargeList using standard KV operations
 * 
 * WARNING: This code is only a first cut and not extensively tested
 * @author peter
 *
 */
public class LargeList {
	public static final String ListElementBinName = "__ListElement";

	private AerospikeClient client;
	private WritePolicy policy;
	private Key key;
	private Value binName;
	private String binNameString;

	public LargeList(AerospikeClient client, WritePolicy policy, Key key, String binName)
	{
		this.client = client;
		this.policy = policy;
		this.key = key;
		this.binName = Value.get(binName);
		this.binNameString = this.binName.toString ();
	}

	private List<byte[]> subRecordList(Key key){
		Record record = client.get (this.policy, key, binNameString);
		if (record != null) {
			return (List<byte[]>)record.getValue (binNameString);
		}
		return null;
	}

	private Key makeSubKey(Value value) {
		Key subKey;
		String valueString;
		if (value instanceof Value.MapValue) {

			Map map = (Map) value.getObject();
			valueString = map.get("key").toString();

		} else {

			valueString = value.toString ();

		}
		String subKeyString = String.format ("%s::%s", this.key.userKey.toString (), valueString);
		subKey = new Key (this.key.namespace, this.key.setName, subKeyString);
		return subKey;
	}

	private Key[] makeSubKeys(List<Value> values) {
		Key[] keys = new Key[values.size()];
		int index = 0;
		for (Value value : values) {
			keys [index] = makeSubKey (value);
			index++;
		}
		return keys;
	}

	private List<byte[]> getDigestList(){
		Record topRecord = client.get (this.policy, this.key, this.binNameString);
		if (topRecord == null)
			return new ArrayList<byte[]> ();
			List<byte[]> digestList = (List<byte[]>) topRecord.getValue (this.binNameString);
			if (digestList == null)
				return new ArrayList<byte[]> ();
				return digestList;
	}

	public void add(Value value)
	{

		Key subKey = makeSubKey (value);

		client.put (this.policy, subKey, new Bin (ListElementBinName, value));

		// add the digest of the subKey to the CDT List in the Customer record
		client.operate(this.policy, this.key, ListOperation.append(this.binNameString, Value.get(subKey.digest)));

	}
	public void add(List<Value> items)
	{
		for(Value Value : items){
			this.add(Value);
		}
	}

	public void add(Value... items)
	{
		for(Value Value : items){
			this.add(Value);
		}
	}
	public void update(Value value)
	{

		if (size () == 0) {
			add (value);
		} else {
			Key subKey = makeSubKey (value);
			client.put (this.policy, subKey, new Bin (ListElementBinName, value));
		}
	}
	public void update(Value... values)
	{
		for(Value value : values){
			this.update(value);
		}
	}

	public void update(List<Value> values)
	{
		for(Value value : values){
			this.update(value);
		}

	}

	public void remove(Value value)
	{
		Key subKey = makeSubKey (value);
		List<byte[]> digestList = getDigestList ();
		int index = digestList.indexOf (subKey.digest);
		client.delete (this.policy, subKey);
		client.operate(this.policy, this.key, ListOperation.remove(this.binNameString, index));

	}

	public void remove(List<Value> values)
	{
		Key[] keys = makeSubKeys (values);
		List<byte[]> digestList = getDigestList ();

		//		int startIndex = digestList.IndexOf (subKey.digest);
		//		int count = values.Count;
		//		foreach (Key key in keys){
		//			
		//			client.Delete (this.policy, key);
		//		}
		//		client.Operate(this.policy, this.key, ListOperation.Remove(this.binNameString, startIndex, count));

		for (Key key : keys){

			client.delete (this.policy, key);
			digestList.remove (key.digest);

		}

		client.put (this.policy, this.key, new Bin (this.binNameString, digestList));

	}

	public int remove(Value begin, Value end)
	{
		List<byte[]> digestList = getDigestList ();
		Key beginKey = makeSubKey (begin);
		Key endKey = makeSubKey (end);
		int start = digestList.indexOf (beginKey.digest);
		int stop = digestList.indexOf (endKey.digest);
		int count = stop - start + 1;;
		for (int i = start; i < stop; i++){
			Key subKey = new Key (this.key.namespace, (byte[])digestList.get(i), null, null);
			client.delete (this.policy, subKey);
		}
		client.operate(this.policy, this.key, ListOperation.removeRange(this.binNameString, start, count));
		return count;
	}
	public boolean exists(Value keyValue)
	{
		Key subKey = makeSubKey (keyValue);
		return client.exists (this.policy, subKey);

	}

	public List<Boolean> exists (List<Value> keyValues)
	{
		List<Boolean> target = new ArrayList<Boolean> ();
		for (Object value : keyValues) {
			target.add (exists (Value.get(value)));
		}
		return target;

	}

	public List<?> find(Value value)
	{
		Key subKey = makeSubKey (value);
		Record record = client.get (this.policy, subKey, ListElementBinName);
		if (record != null) {
			final Object result = record.getValue (ListElementBinName);
			return new ArrayList<Object> (){{ add(result); }};
		} else {
			return null;
		}

	}

	private List<?> get(List<byte[]> digestList, int start, int stop){
		List<Object> results = new ArrayList<Object> ();

		for (int i = start; i < stop; i++) {
			Key subKey = new Key (this.key.namespace, (byte[])digestList.get(i), null, null);
			Record record = client.get (this.policy, subKey, ListElementBinName);
			Object result = record.getValue (ListElementBinName);
			results.add (result);
		}
		return results;
	}

	public List<?> findFrom (Value begin, int count)
	{
		List<byte[]> digestList = getDigestList ();
		Key beginKey = makeSubKey (begin);
		int start = digestList.indexOf (beginKey.digest);
		int stop = start + count;
		return get (digestList, start, stop);
	}
	//
	//	public IList FindFrom(Value begin, int count, string filterModule, string filterName, params Value[] filterArgs)
	//	{
	//		return (IList)client.Execute(policy, key, PackageName, "find_from", binName, begin, Value.Get(count), Value.Get(filterModule), Value.Get(filterName), Value.Get(filterArgs));
	//	}

	public List<?> range(Value begin, Value end)
	{
		List<byte[]> digestList = getDigestList ();
		Key beginKey = makeSubKey (begin);
		Key endKey = makeSubKey (end);
		int start = digestList.indexOf (beginKey.digest);
		int stop = digestList.indexOf (endKey.digest);
		return get (digestList, start, stop);
	}

	//	public IList Range(Value begin, Value end, int count)
	//	{
	//		return (IList)client.Execute(policy, key, PackageName, "find_range", binName, begin, end, Value.Get(count));
	//	}
	//
	//	public IList Range(Value begin, Value end, string filterModule, string filterName, params Value[] filterArgs)
	//	{
	//		return (IList)client.Execute(policy, key, PackageName, "range", binName, begin, end, Value.Get(filterModule), Value.Get(filterModule), Value.Get(filterArgs));
	//	}
	//
	//	public IList Range(Value begin, Value end, int count, string filterModule, string filterName, params Value[] filterArgs)
	//	{
	//		return (IList)client.Execute(policy, key, PackageName, "find_range", binName, begin, end, Value.Get(count), Value.Get(filterModule), Value.Get(filterName), Value.Get(filterArgs));
	//	}

	public List<?> scan()
	{

		List<byte[]> digestList = getDigestList ();
		if (digestList != null || digestList.size() > 0) {
			List<Value> results = new ArrayList<Value> ();
			for (byte[] digest : digestList) {
				Key subKey = new Key (this.key.namespace, digest, null, null);
				Record record = client.get (this.policy, subKey, this.binNameString);
				results.add ((Value) record.getValue (ListElementBinName));
			}
			return results;

		}
		return null;
	}

	//	public IList Filter(string filterModule, string filterName, params Value[] filterArgs)
	//	{
	//		return (IList)client.Execute(policy, key, PackageName, "filter", binName, Value.AsNull, Value.Get(filterModule), Value.Get(filterName), Value.Get(filterArgs));
	//	}

	public void destroy ()
	{
		List<byte[]> digestList = getDigestList ();

		client.put (this.policy, this.key, Bin.asNull(this.binNameString));

		for (byte[] digest : digestList) {
			Key subKey = new Key (this.key.namespace, digest, null, null);
			client.delete (this.policy, subKey);
		}


	}

	public int size()
	{
		Record record = client.operate(this.policy, this.key, ListOperation.size(this.binNameString));
		return record.getInt ("Count");
	}

	public Map<?,?> GetConfig()
	{
		return null;  // No config because its NotSupportedException AsyncNode LDT
	}

	public void SetPageSize(int pageSize)
	{
		// do nothing
	}

}

File: csharp/AerospikeHelper/AerospikeHelper/Collections/LargeList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Aerospike.Client;

namespace Aerospike.Helper.Collections
{
	/// <summary>
	/// Create and manage a list within related records.
	/// This class is a replacement for the LargeList LDT
	/// </summary>

	public sealed class LargeList
	{

		public const String ListElementBinName = "__ListElement";

		private readonly AerospikeClient client;
		private readonly WritePolicy policy;
		private readonly Key key;
		private readonly Value binName;
		private readonly String binNameString;

		/// <summary>
		/// Initialize large list operator.
		/// </summary>
		/// <param name="client">client</param>
		/// <param name="policy">generic configuration parameters, pass in null for defaults</param>
		/// <param name="key">unique record identifier</param>
		/// <param name="binName">bin name</param>
		public LargeList(AerospikeClient client, WritePolicy policy, Key key, string binName)
		{
			this.client = client;
			this.policy = policy;
			this.key = key;
			this.binName = Value.Get(binName);
			this.binNameString = this.binName.ToString ();
		}

		private List<byte[]> SubRecordList(Key key){
			Record record = client.Get (this.policy, key, binNameString);
			if (record != null) {
				return (List<byte[]>)record.GetList (binNameString);
			}
			return null;
		}

		private Key MakeSubKey(Value value) {
			Key subKey;
			String valueString;
			if (value is Value.MapValue) {

				IDictionary map = (IDictionary) value.Object;
				valueString = map ["key"].ToString();

			} else {
				
				valueString = value.ToString ();

			}
			string subKeyString = String.Format ("{0}::{1}", this.key.userKey.ToString (), valueString);
			subKey = new Key (this.key.ns, this.key.setName, subKeyString);
			return subKey;
		}
		private Key[] MakeSubKeys(IList<Value> values) {
			Key[] keys = new Key[values.Count];
			int index = 0;
			foreach (Value value in values) {
				keys [index] = MakeSubKey (value);
				index++;
			}
			return keys;
		}

		private IList GetDigestList(){
			Record topRecord = client.Get (this.policy, this.key, this.binNameString);
			if (topRecord == null)
				return new List<byte[]> ();
			IList digestList = topRecord.GetList (this.binNameString);
			if (digestList == null)
				return new List<byte[]> ();
			return digestList;
		}
		/// <summary>
		/// Add value to list.  Fail if value's key exists and list is configured for unique keys.
		/// If value is a map, the key is identified by "key" entry.  Otherwise, the value is the key.
		/// If large list does not exist, create it.
		/// </summary>
		/// <param name="value">value to add</param>

		public void Add(Value value)
		{
				
			Key subKey = MakeSubKey (value);

			client.Put (this.policy, subKey, new Bin (ListElementBinName, value));

			// add the digest of the subKey to the CDT List in the Customer record
			client.Operate(this.policy, this.key, ListOperation.Append(this.binNameString, Value.Get(subKey.digest)));

		}
		/// <summary>
		/// Add values to list.  Fail if a value's key exists and list is configured for unique keys.
		/// If value is a map, the key is identified by "key" entry.  Otherwise, the value is the key.
		/// If large list does not exist, create it.
		/// </summary>
		/// <param name="values">values to add</param>
		public void Add(List<Value> items)
		{
			foreach(Value Value in items){
				this.Add(Value);
			}
		}
		/// <summary>
		/// Add values to list.  Fail if a value's key exists and list is configured for unique keys.
		/// If value is a map, the key is identified by "key" entry.  Otherwise, the value is the key.
		/// If large list does not exist, create it.
		/// </summary>
		/// <param name="values">values to add</param>

		public void Add(params Value[] items)
		{
			foreach(Value Value in items){
				this.Add(Value);
			}
		}

		/// <summary>
		/// Update value in list if key exists.  Add value to list if key does not exist.
		/// If value is a map, the key is identified by "key" entry.  Otherwise, the value is the key.
		/// If large list does not exist, create it.
		/// </summary>
		/// <param name="value">value to update</param>
		public void Update(Value value)
		{
			
			if (Size () == 0) {
				Add (value);
			} else {
				Key subKey = MakeSubKey (value);
				client.Put (this.policy, subKey, new Bin (ListElementBinName, value));
			}
		}

		/// <summary>
		/// Update/Add each value in array depending if key exists or not.
		/// If value is a map, the key is identified by "key" entry.  Otherwise, the value is the key.
		/// If large list does not exist, create it.
		/// </summary>
		/// <param name="values">values to update</param>
		public void Update(params Value[] values)
		{
			foreach(Value Value in values){
				this.Update(Value);
			}
		}
			
		/// <summary>
		/// Update/Add each value in values list depending if key exists or not.
		/// If value is a map, the key is identified by "key" entry.  Otherwise, the value is the key.
		/// If large list does not exist, create it.
		/// </summary>
		/// <param name="values">values to update</param>
		public void Update(IList values)
		{
			foreach(Value Value in values){
				this.Update(Value);
			}

		}
			
		/// <summary>
		/// Delete value from list.
		/// </summary>
		/// <param name="value">value to delete</param>
		public void Remove(Value value)
		{
			Key subKey = MakeSubKey (value);
			IList digestList = GetDigestList ();
			int index = digestList.IndexOf (subKey.digest);
			client.Delete (this.policy, subKey);
			client.Operate(this.policy, this.key, ListOperation.Remove(this.binNameString, index));

		}

		/// <summary>
		/// Delete values from list.
		/// </summary>
		/// <param name="values">values to delete</param>
		public void Remove(IList<Value> values)
		{
			Key[] keys = MakeSubKeys (values);
			IList digestList = GetDigestList ();

			foreach (Key key in keys){

				client.Delete (this.policy, key);
				digestList.Remove (key.digest);

			}

			client.Put (this.policy, this.key, new Bin (this.binNameString, digestList));
			
		}

		/// <summary>
		/// Delete values from list between range.  Return count of entries removed.
		/// </summary>
		/// <param name="begin">low value of the range (inclusive)</param>
		/// <param name="end">high value of the range (inclusive)</param>
		public int Remove(Value begin, Value end)
		{
			IList digestList = GetDigestList ();
			Key beginKey = MakeSubKey (begin);
			Key endKey = MakeSubKey (end);
			int start = digestList.IndexOf (beginKey.digest);
			int stop = digestList.IndexOf (endKey.digest);
			int count = stop - start + 1;;
			for (int i = start; i < stop; i++){
				Key subKey = new Key (this.key.ns, (byte[])digestList [i], null, null);
				client.Delete (this.policy, subKey);
			}
			client.Operate(this.policy, this.key, ListOperation.RemoveRange(this.binNameString, start, count));
			return count;
		}

		/// <summary>
		/// Does key value exist?
		/// </summary>
		/// <param name="keyValue">key value to lookup</param>
		public bool Exists(Value keyValue)
		{
			Key subKey = MakeSubKey (keyValue);
			return client.Exists (this.policy, subKey);

		}

		/// <summary>
		/// Do key values exist?  Return list of results in one batch call.
		/// </summary>
		/// <param name="keyValues">key values to lookup</param>
		public IList<bool> Exists (IList keyValues)
		{
			IList<bool> target = new List<bool> ();
			foreach (Object value in keyValues) {
				target.Add (Exists (Value.Get(value)));
			}
			return target;

		}

		/// <summary>
		/// Select values from list.
		/// </summary>
		/// <param name="value">value to select</param>
		public IList Find(Value value)
		{
			Key subKey = MakeSubKey (value);
			Record record = client.Get (this.policy, subKey, ListElementBinName);
			if (record != null) {
				Object result = record.GetValue (ListElementBinName);
				return new List<Object> (){ result };
			} else {
				return null;
			}

		}

		private IList get(IList digestList, int start, int stop){
			List<Object> results = new List<object> ();

			for (int i = start; i < stop; i++) {
				Key subKey = new Key (this.key.ns, (byte[])digestList [i], null, null);
				Record record = client.Get (this.policy, subKey, ListElementBinName);
				Object result = record.GetValue (ListElementBinName);
				results.Add (result);
			}
			return results;
		}

		/// <summary>
		/// Select values from the begin key up to a maximum count.
		/// Supported by server versions >= 3.5.8.
		/// </summary>
		/// <param name="begin">start value (inclusive)</param>
		/// <param name="count">maximum number of values to return</param>
		public IList FindFrom (Value begin, int count)
		{
			IList digestList = GetDigestList ();
			Key beginKey = MakeSubKey (begin);
			int start = digestList.IndexOf (beginKey.digest);
			int stop = start + count;
			return get (digestList, start, stop);
		}

		/// <summary>
		/// Select values from the begin key up to a maximum count after applying Lua filter.
		/// 
		/// THIS METHOD IS NOT IMPLEMENTED - DO NOT USE
		/// 
		/// </summary>
		/// <param name="begin">start value (inclusive)</param>
		/// <param name="count">maximum number of values to return after applying Lua filter</param>
		/// <param name="filterModule">Lua module name which contains filter function</param>
		/// <param name="filterName">Lua function name which applies filter to returned list</param>
		/// <param name="filterArgs">arguments to Lua function name</param>
		public IList FindFrom(Value begin, int count, string filterModule, string filterName, params Value[] filterArgs)
		{
			throw new NotImplementedException ();
		}

		/// <summary>
		/// Select range of values from list.
		/// </summary>
		/// <param name="begin">begin value inclusive</param>
		/// <param name="end">end value inclusive</param>
		public IList Range(Value begin, Value end)
		{
			IList digestList = GetDigestList ();
			Key beginKey = MakeSubKey (begin);
			Key endKey = MakeSubKey (end);
			int start = digestList.IndexOf (beginKey.digest);
			int stop = digestList.IndexOf (endKey.digest);
			return get (digestList, start, stop);
		}

		/// <summary>
		/// Select range of values from list.
		/// 
		/// THIS METHOD IS NOT IMPLEMENTED - DO NOT USE
		/// 
		/// </summary>
		/// <param name="begin">low value of the range (inclusive)</param>
		/// <param name="end">high value of the range (inclusive)</param>
		/// <param name="count">maximum number of values to return, pass in zero to obtain all values within range</param>
		public IList Range(Value begin, Value end, int count)
		{
			throw new NotImplementedException ();
		}

		/// <summary>
		/// Select range of values from the large list up to a maximum count after applying lua filter.
		/// 
		/// THIS METHOD IS NOT IMPLEMENTED - DO NOT USE
		/// 
		/// </summary>
		/// <param name="begin">low value of the range (inclusive)</param>
		/// <param name="end">high value of the range (inclusive)</param>
		/// <param name="count">maximum number of values to return after applying lua filter. Pass in zero to obtain all values within range.</param>
		/// <param name="filterModule">Lua module name which contains filter function</param>
		/// <param name="filterName">Lua function name which applies filter to returned list</param>
		/// <param name="filterArgs">arguments to Lua function name</param>
		public IList Range(Value begin, Value end, string filterModule, string filterName, params Value[] filterArgs)
		{
			throw new NotImplementedException ();
		}

		/// <summary>
		/// Select range of values from the large list, then apply a Lua filter.
		/// 
		/// THIS METHOD IS NOT IMPLEMENTED - DO NOT USE
		/// 
		/// </summary>
		/// <param name="begin">low value of the range (inclusive)</param>
		/// <param name="end">high value of the range (inclusive)</param>
		/// <param name="filterModule">Lua module name which contains filter function</param>
		/// <param name="filterName">Lua function name which applies filter to returned list</param>
		/// <param name="filterArgs">arguments to Lua function name</param>
		public IList Range(Value begin, Value end, int count, string filterModule, string filterName, params Value[] filterArgs)
		{
			throw new NotImplementedException ();
		}

		/// <summary>
		/// Return all objects in the list.
		/// </summary>
		public IList Scan()
		{
			
			IList digestList = GetDigestList ();
			if (digestList != null || digestList.Count > 0) {
				IList results = new List<Value> ();
				foreach (byte[] digest in digestList) {
					Key subKey = new Key (this.key.ns, digest, null, null);
					Record record = client.Get (this.policy, subKey, this.binNameString);
					results.Add (record.GetValue (ListElementBinName));
				}
				return results;

			}
			return null;
		}

		/// <summary>
		/// Select values from list and apply specified Lua filter.
		/// 
		/// THIS METHOD IS NOT IMPLEMENTED - DO NOT USE
		/// 
		/// </summary>
		/// <param name="filterModule">Lua module name which contains filter function</param>
		/// <param name="filterName">Lua function name which applies filter to returned list</param>
		/// <param name="filterArgs">arguments to Lua function name</param>
		public IList Filter(string filterModule, string filterName, params Value[] filterArgs)
		{
			throw new NotImplementedException ();
		}

		/// <summary>
		/// Delete bin containing the list.
		/// </summary>
		public void Destroy ()
		{
			IList digestList = GetDigestList ();

			client.Put (this.policy, this.key, Bin.AsNull(this.binNameString));
						
			foreach (byte[] digest in digestList) {
				Key subKey = new Key (this.key.ns, digest, null, null);
				client.Delete (this.policy, subKey);
			}


		}

		/// <summary>
		/// Return size of list.
		/// </summary>
		public int Size()
		{
			Record record = client.Operate(this.policy, this.key, ListOperation.Size(this.binNameString));
			return record.GetInt ("Count");
		}

		/// <summary>
		/// Return map of list configuration parameters.
		/// 
		/// THIS METHOD IS NOT IMPLEMENTED - DO NOT USE
		/// 
		/// </summary>
		public IDictionary GetConfig()
		{
			throw new NotImplementedException ();
		}

		/// <summary>
		/// Set LDT page size. 
		/// 
		/// THIS METHOD IS NOT IMPLEMENTED - DO NOT USE
		/// 
		/// </summary>
		/// <param name="pageSize">page size in bytes</param>
		public void SetPageSize(int pageSize)
		{
			throw new NotImplementedException ();
		}

	}
}

